spec/app/ics-028-cross-chain-validation/technical_specification.md

```diff
@@ -1285,9 +1705,237 @@ function UnbondMaturePackets() {
+    infractionHeight = initH[chainId]
+  }
+  else {
+    infractionHeight = VSCtoH[packet.data.vscId]
+  }
+
+  // request the Slashing module to slash the validator
+  slashingKeeper.Slash(
+    packet.data.valAddress, 
+    infractionHeight, 
+    packet.data.valPower, 
+    packet.data.slashFactor))
+
+  // request the Slashing module to jail the validator
+  timestamp = currentTimestamp() + data.jailTime
+  slashingKeeper.JailUntil(packet.data.valAddress, timestamp)
+
+  // add validator to list of slash requests for chainId
+  slashRequests[chainId].Append(packet.data.valAddress)
+
+  return SlashPacketSuccess
+}
+```
+- Initiator: 
+  - The `onRecvPacket()` method.
+- Expected precondition:
+  - The IBC module on the provider chain received a `SlashPacket` on a channel owned by the provider CCV module.
+- Expected postcondition:
+  - If the channel the packet was sent on is not an established CCV channel, then
+    - the channel status is set to `INVALID`;
+    - the channel closing handshake is initiated;
+    - an error acknowledgment is returned.
+  - Otherwise,
+    - if `packet.data.vscId == 0`, `infractionHeight` is set to `initH[chainId]`, i.e., the height when the first VSC was provided to this consumer chain;
+    - otherwise, `infractionHeight` is set to `VSCtoH[packet.data.vscId]`, i.e., the height at which the voting power was last updated by the validator updates in the VSC with ID `packet.data.vscId`;
+    - a request is made to the Slashing module to slash the validator with address `packet.data.valAddress` for misbehaving at height `infractionHeight`;
+    - a request is made to the Slashing module to jail the validator with address `packet.data.valAddress` for a period `data.jailTime`;
+    - the validator's address `packet.data.valAddress` is added to the list of slash requests from this `chainId`;
+    - a successful acknowledgment is returned.
+- Error condition:
+  - None.
+
+---
+
+<!-- omit in toc -->
+#### **[CCV-CCF-ACKSLASH.1]**
+```typescript
+// CCF: Consumer Chain Function
+function onAcknowledgeSlashPacket(packet: Packet, ack: bytes) {
+  // slash request fail, i.e., ack == SlashPacketError, 
+  // only if the SlashPacket was sent on a channel 
+  // other than the established CCV channel;
+  // that should never happen,
+  // see SendSlashRequest() and SendPendingSlashRequests()
+  abortSystemUnless(ack != SlashPacketError)
+}
+```
+- Initiator: 
+  - The `onAcknowledgePacket()` method.
+- Expected precondition:
+  - The IBC module on the consumer chain received an acknowledgement of a `SlashPacket` on a channel owned by the consumer CCV module.
+- Expected postcondition:
+  - The state is not changed.
+- Error condition:
+  - The acknowledgement is `SlashPacketError`.
+
+<!-- omit in toc -->
+#### **[CCV-CCF-TOSLASH.1]**
+```typescript
+// CCF: Consumer Chain Function
+function onTimeoutSlashPacket(packet Packet) {
+  channelStatus = INVALID
+
+  // TODO What do we do here? 
```

Comment:
Is this related to https://github.com/cosmos/ibc/issues/669? Is this something we need to figure out immediately?

Reply:
Yes, it is related to https://github.com/cosmos/ibc/issues/669. There are opened issues for all the TODOs; I keep them also in the spec to emphasized that's something we are working on.

Reply:
Great. In future project we might link the issue in the files for context. If that's too much overhead don't worry.